File: src/main/java/macro/library/external/OpenLibrary.java
package macro.library.external;

import macro.library.Util;
import macro.library.author.Author;
import macro.library.book.Book;
import macro.library.book.Isbn;
import macro.library.database.AuthorTable;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Macro303 on 2019-Oct-22
 */
public abstract class OpenLibrary {
	@NotNull
	private static final Logger LOGGER = LogManager.getLogger(OpenLibrary.class);
	@NotNull
	private static final String URL = "http://openlibrary.org/api/books";

	@Nullable
	public static Book searchBook(@NotNull Isbn isbn) {
		var url = URL + "?bibkeys=ISBN:" + isbn + "&format=json&jscmd=data";
		var request = Util.httpRequest(url);
		if (request == null)
			return null;
		JSONObject response = request.getObject();
		var bookObj = response.optJSONObject("ISBN:" + isbn);
		if (bookObj == null)
			return null;
		var title = bookObj.getString("title");
		var subtitle = bookObj.optString("subtitle");
		if (subtitle.isBlank())
			subtitle = null;
		var publishers = bookObj.optJSONArray("publishers");
		if (publishers == null)
			publishers = new JSONArray();
		StringBuilder publisherStr = new StringBuilder();
		for (var publisher : publishers) {
			if (publisherStr.length() != 0)
				publisherStr.append(";");
			publisherStr.append(((JSONObject) publisher).getString("name"));
		}
		return new Book(isbn, title, subtitle, publisherStr.toString());
	}

	@NotNull
	public static List<Author> searchAuthors(@NotNull Isbn isbn) {
		var authorList = new ArrayList<Author>();
		var url = URL + "?bibkeys=ISBN:" + isbn + "&format=json&jscmd=data";
		var request = Util.httpRequest(url);
		if (request == null)
			return authorList;
		JSONObject response = request.getObject();
		var bookObj = response.optJSONObject("ISBN:" + isbn);
		if (bookObj == null)
			return authorList;
		var authors = bookObj.optJSONArray("authors");
		if (authors == null)
			authors = new JSONArray();
		for (var authorName : authors) {
			var name = ((JSONObject) authorName).getString("name");
			var author = Author.parseName(name);
			var found = AuthorTable.INSTANCE.select(author.getFirstName(), author.getLastName());
			if (found == null)
				author = author.add();
			authorList.add(author);
		}
		return authorList;
	}
}

File: src/main/kotlin/macro/library/external/OpenLibrary.kt
package macro.library.external

import macro.library.Util
import macro.library.book.Book
import macro.library.book.Isbn
import macro.library.book.author.Author
import macro.library.database.AuthorTable
import org.apache.logging.log4j.LogManager
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

/**
 * Created by Macro303 on 2019-Nov-12
 */
object OpenLibrary {
	private val LOGGER = LogManager.getLogger(OpenLibrary::class.java)
	private val URL = "http://openlibrary.org/api/books"

	fun searchBook(isbn: Isbn): Book? {
		val url = "$URL?bibkeys=ISBN:$isbn&format=json&jscmd=data"
		val request = Util.httpRequest(url) ?: return null
		val response = request.getObject()
		val bookObj = response.optJSONObject("ISBN:$isbn") ?: return null
		val title = bookObj.getString("title")
		var subtitle: String? = bookObj.optString("subtitle")
		if (subtitle.isNullOrBlank())
			subtitle = null
		var publishers: JSONArray? = bookObj.optJSONArray("publishers")
		if (publishers == null)
			publishers = JSONArray()
		val publisherStr = StringBuilder()
		for (publisher in publishers) {
			if (publisherStr.isNotEmpty())
				publisherStr.append(";")
			publisherStr.append((publisher as JSONObject).getString("name"))
		}
		val imageSmall = bookObj.getJSONObject("cover").getString("small")
		val imageMedium = bookObj.getJSONObject("cover").getString("medium")
		val imageLarge = bookObj.getJSONObject("cover").getString("large")
		searchAuthors(isbn)
		return Book(
			isbn = isbn,
			title = title,
			subtitle = subtitle,
			publisher = publisherStr.toString(),
			imageSmall = imageSmall,
			imageMedium = imageMedium,
			imageLarge = imageLarge
		)
	}

	fun searchAuthors(isbn: Isbn): List<Author> {
		val authorList = ArrayList<Author>()
		val url = "$URL?bibkeys=ISBN:$isbn&format=json&jscmd=data"
		val request = Util.httpRequest(url) ?: return authorList
		val response = request.getObject()
		val bookObj = response.optJSONObject("ISBN:$isbn") ?: return authorList
		var authors: JSONArray? = bookObj.optJSONArray("authors")
		if (authors == null)
			authors = JSONArray()
		for (authorName in authors) {
			val name = (authorName as JSONObject).getString("name")
			var author = Author.parseName(name)
			val found = AuthorTable.select(author.firstName, author.lastName)
			if (found == null)
				author = author.add()
			authorList.add(author)
		}
		return authorList
	}
}